osf/models/user.py

```diff
@@ -1508,6 +1508,56 @@ def get_node_comment_timestamps(self, target_id):
         default_timestamp = dt.datetime(1970, 1, 1, 12, 0, 0, tzinfo=pytz.utc)
         return self.comments_viewed_timestamp.get(target_id, default_timestamp)
 
+    def gdpr_delete(self):
+        """
+        This function does not remove the user object reference from our database, but it does disable the account and
+        remove identifying in a manner compliant with GDPR guidelines.
+        :return:
+        """
+        from osf.models import PreprintService, AbstractNode
+
+        user_nodes = self.nodes.all()
+        #  Validates the user isn't trying to delete things they deliberately made public.
+        if user_nodes.filter(type='osf.registration').exists():
+            raise ValidationError('You cannot delete this user because they have one or more registrations.')
+
+        if PreprintService.objects.filter(node___contributors=self, ever_public=True).exists():
+            raise ValidationError('You cannot delete this user because they have one or more preprints.')
```

Comment:
Drive-by: this script should handle _deactivating_ users that have preprints and/or registrations.

Reply:
@caseyrollins I think admins can already deactivate users manually if they get this message. I'm not sure it helps to deactivate them automatically. If the user really wants to delete themselves and they have to change ownership of a shared project first they won't be able to do it if they're deactivated.